### PR TITLE
SVG and raster images can be embedded in project

### LINKED ISSUE #2814: [FEATURE] Allow embedding SVG images in projects in base64 format
Original commit: https://github.com/qgis/QGIS/commit/19f8a0078d2c90036178db71b8e9e1de547fee93 by nyalldawson

Allows SVG images for symbology, labels, etc to be directly embedded
inside a project file (or QML style, or QPT print template!)
by encoding the svg as a standard base64 string.

Sponsored by SMEC/SJ

### LINKED ISSUE #2815: [FEATURE] Better UI for embedding SVG files
Original commit: https://github.com/qgis/QGIS/commit/1a7ade7575ce30fd7f17cdd9cf102a5b41b14e08 by nyalldawson

Adds a common widget for SVG sources, with a tool button with
some handy options:
- select file (old behaviour), pick a file from disk
- embed file (pick a file from disk, is embedded into project/symbol)
- extract embedded file (for embedded files, allows you to save these
back to a disk based svg file)
- from url (opens a dialog prompting for a url, exposing
the previously hidden functionality that svgs can be retrieved
from a remote url (eg github))

Sponsored by SMEC/SJ

### LINKED ISSUE #3216: [FEATURE] Raster image marker can have images set to remote URLS
or as embedded files
Original commit: https://github.com/qgis/QGIS/commit/51b2ce4b6c1a91e8cd3cdc41d8f94c14a5035067 by nyalldawson

Just like the SVG marker, raster markers can now be set to use
HTTP(s) URLs or have files directly embedded inside the
symbol itself.

### LINKED ISSUE #3219: [FEATURE] Raster fill can have images set to remote URLs
or as embedded files.
Original commit: https://github.com/qgis/QGIS/commit/0983f5206396208437c737f2465c0f2b8bcb3e0e by nirvn

Just like the SVG and raster image marker, raster fill can now
be set to use HTTP(s) URLs or have files directly embedded
inside the symbol itself.

### PR DESCRIPTION
**Pull request by DelazJ**
Fix #3219, fix #3216, fix #2815, fix #2814

### CODE REVIEW

**Comment by DelazJ** [comment]
@andreasneumann @rduivenvoorde you were among [the happiest people](https://github.com/qgis/QGIS/pull/7433) :smile:  when the feature was committed. Can you check whether there's some valuable information missing or that i'd have misunderstood, please?
@nyalldawson the commit of #2814 mentions that svg can be saved in qpt files but i can't find a feature in layout that has this new widget for file selection. What did I miss?

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 1]
```diff
@@ -118,11 +118,31 @@ General Settings
...
-symbols`. These SVG files are then available to symbolize features or
-decorate your map composition.
+symbols`. These SVG files are then available to symbolize or label the features
+or decorate your map composition.
+
+When using an SVG file in a symbol or a label, QGIS allows you to:
+
+* load the file from the disk: the file is identified through the file path and
+  QGIS needs to resolve the path in order to display the corresponding image
+* load the file from a remote URL: as above, on the success of the connection
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
* load the file from a remote URL: as above, the image will only be loaded on successful retrieval of the remote resource
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 2]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+  depends the image retrieval
+* embed the SVG file into the item: the file is saved as a base64 string within
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]: "the file is saved as a base64 string within"

Is this relevant for users? It's not exposed anywhere, and could potentially be changed in future.

Comment (nyalldawson) [reply source/docs/user_manual/introduction/qgis_configuration.rst]: Maybe just "the file is embedded inside the current project, style database, or print layout template."

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 3]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+  the symbol or label, and also in the project, an associated style or a print
+  layout template file. THe SVG file is then always rendered as part of the item.
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
  layout template file. The SVG file is then always rendered as part of the item.
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 4]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+  A convenient way to share projects with custom SVG symbols.
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggeston
  This is a convenient way to create self-contained projects with custom SVG symbols which can be easily shared amongst different users and installations of QGIS.
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 5]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+
+  It's as well possible to extract the embedded SVG file from a symbol or label
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
  It is also possible to extract the embedded SVG file from a symbol or label
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 6]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+  and save it on disk.
+
+.. note:: The abovementioned ways of loading/storing an SVG file in a project
```
Comment (nyalldawson) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
.. note:: The above mentioned options for embedding an SVG file in a project
```

Comment (DelazJ) [reply source/docs/user_manual/introduction/qgis_configuration.rst]: It's not only the embedding option I'm referring to: the options to load from file disk and URL are intended to be included, too.

**Comment by DelazJ** [comment]
@nyalldawson Thanks for the quick review. Any element regarding the print layout case?

**Comment by nyalldawson** [comment]
It's not currently supported in print layouts

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 7]
```diff
@@ -118,11 +118,31 @@ General Settings
...
 
 Add or Remove :guilabel:`Path(s) to search for Scalable Vector Graphic (SVG)
-symbols`. These SVG files are then available to symbolize features or
-decorate your map composition.
+symbols`. These SVG files are then available to symbolize or label the features
+or decorate your map composition.
+
+When using an SVG file in a symbol or a label, QGIS allows you to:
+
+* load the file from the disk: the file is identified through the file path and
```
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
* load the file from the file system: the file is identified through the file path and
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 8]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+  QGIS needs to resolve the path in order to display the corresponding image
+* load the file from a remote URL: as above, on the success of the connection
+  depends the image retrieval
```
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
```

File: source/docs/user_manual/introduction/qgis_configuration.rst (hunk 2 above)
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
* embed the SVG file into the item: the file is embedded inside
```

File: source/docs/user_manual/introduction/qgis_configuration.rst [hunk 9]
```diff
@@ -118,11 +118,31 @@ General Settings
...
+* embed the SVG file into the item: the file is saved as a base64 string within
+  the symbol or label, and also in the project, an associated style or a print
```
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
  the current project, style database, or print layout template.
```

File: source/docs/user_manual/introduction/qgis_configuration.rst (hunk 3 above)
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
```

File: source/docs/user_manual/introduction/qgis_configuration.rst (hunk 4 above)
Comment (havatv) [inline source/docs/user_manual/introduction/qgis_configuration.rst]:
```suggestion
  This is a convenient way to create self-contained projects with custom SVG symbols
  which can be easily shared amongst different users and installations of QGIS.
```

File: source/docs/user_manual/introduction/qgis_configuration.rst (hunk 3 above)
Comment (DelazJ) [reply source/docs/user_manual/introduction/qgis_configuration.rst]: @havatv you'd remove the second sentence?

**Pull request merged**